service/worker/archiver/pump.go

```diff
@@ -0,0 +1,141 @@
+	PumpResult struct {
+		PumpedHashes          []uint64
+		UnhandledCarryover    []ArchiveRequest
+		TimeoutWithoutSignals bool
+	}
+
+	pump struct {
+		ctx           workflow.Context
+		logger        bark.Logger
+		metricsClient metrics.Client
+		carryover     []ArchiveRequest
+		timeout       time.Duration
+		requestLimit  int
+		requestCh     workflow.Channel
+		signalCh      workflow.Channel
+	}
+)
+
+// NewPump returns a new Pump
+func NewPump(
+	ctx workflow.Context,
+	logger bark.Logger,
+	metricsClient metrics.Client,
+	carryover []ArchiveRequest,
+	timeout time.Duration,
+	requestLimit int,
+	requestCh workflow.Channel,
+	signalCh workflow.Channel,
+) Pump {
+	return &pump{
+		ctx:           ctx,
+		logger:        logger,
+		metricsClient: metricsClient,
+		carryover:     carryover,
+		timeout:       timeout,
+		requestLimit:  requestLimit,
+		requestCh:     requestCh,
+		signalCh:      signalCh,
+	}
+}
+
+// Run pumps requests into request channel.
+// Blocks until either timout occurs or request limit is satisfied.
+// Returns a PumpResult which contains a summary of what was pumped.
+// Upon returning request channel is closed.
+func (p *pump) Run() PumpResult {
+	defer func() {
+		p.requestCh.Close()
+	}()
+
+	sw := p.metricsClient.StartTimer(metrics.ArchiverPumpScope, metrics.CadenceLatency)
+	defer sw.Stop()
+
+	carryoverBoundIndex := len(p.carryover)
+	if carryoverBoundIndex > p.requestLimit {
+		carryoverBoundIndex = p.requestLimit
+	}
+	var unhandledCarryover []ArchiveRequest
+	for i := carryoverBoundIndex; i < len(p.carryover); i++ {
+		unhandledCarryover = append(unhandledCarryover, p.carryover[i])
+	}
+	p.metricsClient.UpdateGauge(metrics.ArchiverPumpScope, metrics.ArchiverBacklogSizeGauge, float64(len(unhandledCarryover)))
+	pumpResult := PumpResult{
+		UnhandledCarryover: unhandledCarryover,
+	}
+	for i := 0; i < carryoverBoundIndex; i++ {
+		request := p.carryover[i]
+		p.requestCh.Send(p.ctx, request)
+		pumpResult.PumpedHashes = append(pumpResult.PumpedHashes, hashArchiveRequest(request))
+	}
+	if len(pumpResult.PumpedHashes) == p.requestLimit {
+		return pumpResult
+	}
+	selector := workflow.NewSelector(p.ctx)
+	finished := false
+	selector.AddFuture(workflow.NewTimer(p.ctx, p.timeout), func(_ workflow.Future) {
+		p.metricsClient.IncCounter(metrics.ArchiverPumpScope, metrics.ArchiverPumpTimeoutCount)
+		if len(p.carryover) == len(pumpResult.PumpedHashes) {
+			p.metricsClient.IncCounter(metrics.ArchiverPumpScope, metrics.ArchiverPumpTimeoutWithoutSignalsCount)
+			pumpResult.TimeoutWithoutSignals = true
+		}
+		finished = true
+	})
+	selector.AddReceive(p.signalCh, func(ch workflow.Channel, more bool) {
+		if !more {
+			p.logger.Error("signal channel channel closed unexpectedly")
+			p.metricsClient.IncCounter(metrics.ArchiverPumpScope, metrics.ArchiverPumpSignalChannelClosedCount)
+			finished = true
+			return
+		}
+		var request ArchiveRequest
+		ch.Receive(p.ctx, &request)
```

Comment:
Let's add some handling if we are unable to deserialize payload into ArchiveRequest.

Reply:
The Receive API does not lend itself to this error handling. Lets chat about best approach here.